plugins/aws-s3-storage/src/index.ts

```diff
@@ -227,6 +195,68 @@ export default class S3Database implements IPluginStorage<S3Config> {
     return this._localData as LocalStorage;
   }
 
+  private async _getPackageList(): Promise<any> {
+    const logger = this.logger;
+    const { bucket, keyPrefix } = this.config;
+    const listPath = keyPrefix.endsWith('/') ? keyPrefix : '/';
+    logger.debug(
+      { keyPrefix, bucket, listPath },
+      's3: [_getPackageList] bucket: @{bucket} prefix: @{keyPrefix} listPath: @{listPath}'
+    );
+    const lister = s3ls({ bucket, s3: this.s3 });
+    // Generate tree { packages:[], scopes[] } data
+    function listS3Folder(folder: string): Promise<any> {
+      if (!folder.endsWith('/')) {
+        folder = folder + '/';
+      }
+      return lister.ls(folder).then(data => {
+        const { folders } = data;
+        logger.debug({ folders }, 's3: [_getPackageList.listS3Folder] folders: @{folders}');
+        // folders of root level: ['prefix/package-a/', 'prefix/package-b/', 'prefix/@scope/', ...]
+        // folders of scoped level: [prefix/@scope/package-c/', ...]
+        const tree = {
+          packages: [] as string[],
+          scopes: [] as string[],
+        };
+        folders.forEach(folder => {
+          const segs = folder.split('/');
+          const seg = segs[segs.length - 2];
+          if (seg.startsWith('@')) {
+            tree.scopes.push(seg);
+          } else {
+            tree.packages.push(seg);
+          }
+        });
+        return Promise.resolve(tree);
+      });
+    }
+    return listS3Folder(listPath).then(tree => {
+      if (!tree.scopes.length) {
```

Comment:
Branch is unneeded. If there are no scopes, the `Promise.all()` will resolve immediately and we'll simply return `tree.packages` as we expect.

Reply:
Looks good